apps/datapuller/src/pullers/classes.ts

```diff
@@ -10,6 +10,73 @@ import {
 
 const TERMS_PER_API_BATCH = 4;
 
+const checkTermHasCatalogData = async (
+  year: number,
+  semester: string,
+  academicCareerCode: string
+): Promise<boolean> => {
+  const termExists = await TermModel.exists({
+    name: `${year} ${semester}`,
+    academicCareerCode,
+  });
+
+  if (!termExists) return false;
+
+  const classExists = await ClassModel.exists({
+    year,
+    semester,
+  });
+
+  return !!classExists;
+};
+
+// Note: We scan ALL terms in the database rather than just the affected ones.
+// This is intentional to ensure data consistency, as class data may be modified
+// by sources other than the datapuller. Since the query is fast, the overhead
+// is minimal. Future optimization could track only affected terms if needed.
+const updateTermsCatalogDataFlags = async (log: any) => {
+  log.trace("Updating hasCatalogData flags for all terms...");
+
+  const allTerms = await TermModel.find({})
+    .select({ name: 1, academicCareerCode: 1 })
+    .lean();
+
+  const BATCH_SIZE = 50;
+  const bulkOps = [];
+
+  for (let i = 0; i < allTerms.length; i += BATCH_SIZE) {
+    const batch = allTerms.slice(i, i + BATCH_SIZE);
+    const batchOps = await Promise.all(
+      batch.map(async (term) => {
+        const [year, semester] = term.name.split(" ");
+        const hasCatalogData = await checkTermHasCatalogData(
+          parseInt(year),
+          semester,
+          term.academicCareerCode
+        );
```

Comment:
Why not just do:

```ts
  const allTerms = await TermModel.find({})
    .select({ _id: 1, id: 1 })
    .lean();
```

And then pass in `id` into `checkTermHasCatalogData`. Then, you can remove the extra term query in the in `checkTermHasCatalogData` and do:

```ts
const count = await ClassModel.countDocuments({ termId });
```

Then, for the update query:

```ts
updateOne: {
  filter: { _id },
  update: { $set: { hasCatalogData } },
}
```